libmoep-1.1/examples/moepeval.1
.TH MOEPEVAL 1 2014-5-5 "moep80211" "MOEP80211 Manual"
.SH NAME
moepeval \- a feature evaluator for moep80211
.SH SYNOPSIS
.B moepeval
.RI [ OPTION "]... " IF1 " " IF2
.SH DESCRIPTION
This program evaluates different driver and hardware features. Therefore it uses
the two radio interfaces IF1 and IF2.
.TP
.BR "-?" ", " "--help"
Give a help list
.TP
.BR "--usage"
Give a short usage message
.TP
.BR "-V" ", " "--version"
Print program version
.P
Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.
.SH BUGS
Report bugs to <leclaire@in.tum.de>.


mdk3/CHANGELOG
MDK3 Changelog

version 6
* Amok Mode now works on Ad-Hoc and MANET networks (WARNING: Clients may not reconnect automatically,
  so they may stay disconnected after the attack stopped!)
* Removed duplicate WPA downgrade in Deauth Mode (sry for the confusion)
* SSID Bruteforce Mode understands 0 and 1 byte placeholders in hidden SSIDs now
* GCC 4.4 support, all warnings and extra warnings fixed
* Whitelists and Blacklists in Amok Mode are re-read periodically every 3 seconds.
  You can use this to dynamically allow or block hosts with scripts.
* A lot of small bugfixes

version 5
* Enhanced MAC-Filter Bruteforce Mode
* Another WDS/WIDS/WIPS Confusion Test
* Amok Mode supports QoS packets
* Michael Countermeasure Exploit (also known as TKIP QoS Exploit)
  can shut down APs using TKIP encryption and QoS Extension with 1 sniffed and 2 injected QoS Data Packets.
* WPA-Downgrade Test - deauthenticates Stations and APs sending WPA encrypted packets.
  With this test you can check if the sysadmin will try setting his network to WEP or disable encryption.
  mdk3 will let WEP and unencrypted clients work, so if the sysadmin simply thinks "WPA is broken" he sure isn't the right one for this job.
  (this can/should be combined with social engineering)

version 4
* Added experimental high speed MAC-Filter Bruteforce Mode

version 3
* Added a channel hopper for Amok Mode
* Added WIDS confusion mode
* fresh & bugfixed osdep included
* Fixed White- and Blacklisting again...

version 2
* More Documentation
* Added -Wall to Makefile to always keep the code clean
* Fixed the Warnings produced by -Wall ;)
* Updated osdep
* Merged some patches from Andy Green to clean up the code
* Poured some holy water all over the code, because mdk3 is used by professionals
* Added intelligent Authentication DoS mode
* Fixed White- and Blacklist function in Amok mode

version 1
* NOW USING OSDEP INJECTION from aircrack-ng project
  -> mdk2 should now run on LINUX and FREEBSD (and soon more)
* Started writing some docs (now that mdk3 may soon run on Windows, the kids may need it...)


MDK2 version history:
version 37 (has never been released, went directly to mdk3)
* Better Madwifi-ng handling
* Blacklist mode fixed

version 36
* Blacklist support for Deauth Mode
* Fixed some minor gcc warnings
* Advanced FakeAP mode. Reads SSIDs and MACs from file.
  See example file.

version 35
* telek0miker seems to have his EAPOL-Start test fixed
* Help screen is now splitted into several sexy parts.

version 34
* Compilation fix for PPC architecture
* Le_Vert added all copyright and license informations. (thx)
* Rewrite Makefile.

version 33
* Added telek0miker's EAPOL-Logoff Test (thanks again)

version 32
* Added telek0miker's EAPOL-Start flood (thanx!)

version 31
* fixed race condition in bruteforce mode

version 30
AKA "Hell, I don't know how to do the versioning!"
* Just counting version numbers now, no beta and dev versions
* Added beacon flood variety patch from ducttape (THANKS!)
* Ace's SSID Bruteforcer works now

beta version 3
AKA "Zuo Meng Hua, I love you!"-Edition
* Used thefkboss' "motherfucker" version
* Added Whitelist support for deauth mode
* Updated TODO list, looks good to have more "DONE" in there ;)
* Cleaned up a little
* Stabilized and cleaned SSID Bruteforcer
* Added Ace's SSID Bruteforcer - HUGE THANKS! (Doesnt work yet...)
* Added Hirte's TKIP Shutdown Exploit - THANK YOU!
* Makefile fixed

development version 0028
AKA BETA 2
* cleaned up code, removed gcc warnings
* fixed -e/-s confusion in probe mode
* fixed signedness bug in beacon mode display
* now using Makefile from Zero_Chaos (thx!)
* implemented Ace's SSID Bruteforce (thx!)
* probe checks works on rt2570 using my 1.4.0 driver (required for SSID Bruteforce)

development version 0027
* speeding up tests which need to sniff
* Added option to use valid MAC adresses from OUI database

development version 0026
* Added SSID wordlist mode (thanks to moongray)
* Added SSID lenght checkings and warnings
* Total rewrite of Deauth-Amok-Mode, now it's even more aggressive ;)

development version 0024
* fixed header files for newer distributions (SuSE 10.1, FC5, maybe others)
* fixed very stupid bug causing segfault in some cases
* Added some more options to Fake AP mode (thanks to Antragon)

beta version 1 - had the feeling to make one... ;)
* added fully automatic deauthentication/disassociation mode (Amok-Mode!)

development version 00xx
* added probe mode
* probe check does not work on rt2570, need to investigate why

development version 0019
* tweaked auth dos to react faster in conditions with low traffic
* additional information and out of range warning in auth DoS checker
* Beacon Flood mode now supports channel hopping

development version 0016
* auth dos mode more fixup
* auth dos checker is now kind of mature (perhaps needs tweaking)
* cleaning up code, removing unused variables
* adding a lot of comments to make easier to understand

development version 0014
* auth dos mode fixed
* auth dos checker working very unreliable
* New method to generate random SSIDs

development version 0012
* basic beacon mode functional
* very basic and buggy auth dos mode without check for success
* this code was mostly copied from old mdk version

preparing version 0000
* preparing copied functions from aireplay


libmoep-1.1/examples/moepeval.c
/*
 * Copyright 2013, 2014		Maurice Leclaire <leclaire@in.tum.de>
 *				Stephan M. Guenther <moepi@moepi.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * See COPYING for more details.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <argp.h>
#include <endian.h>

#include <arpa/inet.h>

#include <net/ethernet.h>

#include <netinet/in.h>

#include <sys/eventfd.h>

#include <moep80211/system.h>
#include <moep80211/types.h>
#include <moep80211/ieee80211_addr.h>

#include <moep80211/modules/moep80211.h>

#include "../src/util.h"


const char *argp_program_version = "moepeval 1.0";
const char *argp_program_bug_address = "<leclaire@in.tum.de>";

static char args_doc[] = "IF1 IF2";

static char doc[] =
"moepeval - a feature evaluator for moep80211\n\n"
"  IF1                        Radio interface number 1\n"
"  IF2                        Radio interface number 2";

enum fix_args {
	FIX_ARG_IF1 = 0,
	FIX_ARG_IF2 = 1,
	FIX_ARG_CNT
};

static struct argp_option options[] = {
	{}
};

static error_t parse_opt(int key, char *arg, struct argp_state *state);

static struct argp argp = {
	options,
	parse_opt,
	args_doc,
	doc
};


struct arguments {
	char *if1;
	char *if2;
};


#define DEFAULT_MTU 500
#define DEFAULT_CHAN chan_freq[0]

static u64 chan_freq[] = {
	2412,
	2417,
	2422,
	2427,
	2432,
	2437,
	2442,
	2447,
	2452,
	2457,
	2462,
	2467,
	2472,
	2484,

	5180,
	5200,
	5220,
	5240,
	5260,
	5280,
	5300,
	5320,

	5500,
	5520,
	5540,
	5560,
	5580,
	5600,
	5620,
	5640,
	5660,
	5680,
	5700,

	5735,
	5755,
	5775,
	5795,
	5815,
	5835,
	5855,
};

int legacy_rates[] = {
	2, 4, 11, 12, 18, 22, 24, 36, 48, 72, 96, 108
};

static int legacy_idx(int rate)
{
	int i;

	for (i = 0; i < ARRAY_SIZE(legacy_rates); i++) {
		if (legacy_rates[i] == rate)
			return i;
	}
	return -1;
}


static error_t parse_opt(int key, char *arg, struct argp_state *state)
{
	struct arguments *args = state->input;

	switch (key) {
	case ARGP_KEY_ARG:
		switch (state->arg_num) {
		case FIX_ARG_IF1:
			args->if1 = arg;
			break;
		case FIX_ARG_IF2:
			args->if2 = arg;
			break;
		default:
			argp_usage(state);
		}
		break;
	case ARGP_KEY_END:
		if (state->arg_num < FIX_ARG_CNT)
			argp_usage(state);
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}


static int dev_avail(const char *name)
{
	moep_dev_t dev;

	if (!(dev = moep_dev_moep80211_open(name, DEFAULT_CHAN,
					    MOEP80211_CHAN_WIDTH_20_NOHT, 0, 0,
					    DEFAULT_MTU))) {
		fprintf(stderr, "cannot create device on '%s': %s\n", name, strerror(errno));
		return -1;
	}
	moep_dev_close(dev);

	return 0;
}

static int max_mtu(const char *name)
{
	moep_dev_t dev;
	int i, step;

	for (i = DEFAULT_MTU, step = i/2; 1; i += step) {
		if (!step)
			break;
		if (!(dev = moep_dev_moep80211_open(name, DEFAULT_CHAN,
						    MOEP80211_CHAN_WIDTH_20_NOHT,
						    0, 0, i))) {
			i -= step;
			step /= 2;
			continue;
		}
		moep_dev_close(dev);
		printf("\rmax MTU %s: %d", name, i);
		fflush(stdout);
	}
	printf("\n");

	return i;
}

static int avail_channels(const char *name)
{
	moep_dev_t dev;
	int i;
	u64 chan;

	chan = 0;
	printf("available channels %s: ", name);
	for (i = 0; i < ARRAY_SIZE(chan_freq); i++) {
		if (!(dev = moep_dev_moep80211_open(name, chan_freq[i],
						    MOEP80211_CHAN_WIDTH_20_NOHT,
						    0, 0, DEFAULT_MTU))) {
			printf("-");
			fflush(stdout);
			continue;
		}
		moep_dev_close(dev);
		chan |= BIT(i);
		printf("+");
		fflush(stdout);
	}
	printf("\n");

	return chan;
}

static u8 data_rate_work_payload[] = "This is the payload";

struct moep_hdr_pctrl data_rate_hdr_pctrl = {
	.hdr = {
		.type = MOEP_HDR_PCTRL,
		.len = sizeof(struct moep_hdr_pctrl),
	},
	.type = htole16(0),
	.len = htole16(sizeof(data_rate_work_payload)),
};

static u16 legacy_rate_worked;
static u32 mcs_rate_worked;

static void rx_handler_work(moep_dev_t dev, moep_frame_t frame)
{
	struct moep80211_radiotap *radiotap;
	struct moep80211_hdr *hdr;
	struct moep_hdr_pctrl *pctrl;

	if (!(hdr = moep_frame_moep80211_hdr(frame))) {
		moep_frame_destroy(frame);
		return;
	}
	if (hdr->frame_control !=
	    htole16(IEEE80211_FTYPE_DATA | IEEE80211_STYPE_DATA)) {
		moep_frame_destroy(frame);
		return;
	}

	if (!(pctrl = (struct moep_hdr_pctrl *)moep_frame_moep_hdr_ext(frame, MOEP_HDR_PCTRL))) {
		moep_frame_destroy(frame);
		return;
	}
	if (pctrl->hdr.len != sizeof(struct moep_hdr_pctrl)) {
		moep_frame_destroy(frame);
		return;
	}
	if (pctrl->len != htole16(sizeof(data_rate_work_payload))) {
		moep_frame_destroy(frame);
		return;
	}
	if (pctrl->type != htole16(0)) {
		moep_frame_destroy(frame);
		return;
	}

	// TODO validate payload

	if (!(radiotap = moep_frame_radiotap(frame))) {
		moep_frame_destroy(frame);
		return;
	}
	if (radiotap->hdr.it_present & BIT(IEEE80211_RADIOTAP_RATE))
		legacy_rate_worked |= BIT(legacy_idx(radiotap->rate));
	if ((radiotap->hdr.it_present & BIT(IEEE80211_RADIOTAP_MCS)) &&
	    (radiotap->mcs.known & IEEE80211_RADIOTAP_MCS_HAVE_MCS))
		mcs_rate_worked |= BIT(radiotap->mcs.mcs);

	moep_frame_destroy(frame);
}

static void test_data_rates_work(const char *name1, const char *name2, u64 freq)
{
	moep_dev_t dev1, dev2;
	int i, k;
	struct timespec timeout;
	moep_frame_t frame;
	struct moep80211_radiotap *radiotap;
	struct moep80211_hdr *hdr;
	int event_all;

	printf("%s -> %s, %ld MHz: ", name1, name2, freq);

	if (!(dev1 = moep_dev_moep80211_open(name1, freq,
					     MOEP80211_CHAN_WIDTH_20_NOHT, 0, 0,
					     DEFAULT_MTU))) {
		fprintf(stderr, "cannot create device on %s, %s\n", name1, strerror(errno));
		return;
	}
	if (!(dev2 = moep_dev_moep80211_open(name2, freq,
					     MOEP80211_CHAN_WIDTH_20_NOHT, 0, 0,
					     DEFAULT_MTU))) {
		fprintf(stderr, "cannot create device on %s, %s\n", name2, strerror(errno));
		moep_dev_close(dev1);
		return;
	}
	moep_dev_set_rx_handler(dev2, rx_handler_work);
	if ((event_all = eventfd(1, EFD_NONBLOCK | EFD_SEMAPHORE)) < 0) {
		fprintf(stderr, "cannot create eventfd, %s\n", strerror(errno));
		moep_dev_close(dev1);
		moep_dev_close(dev2);
		return;
	}
	moep_dev_set_rx_event(dev2, event_all);

	if (!(frame = moep_dev_frame_create(dev1))) {
		fprintf(stderr, "cannot create frame, %s\n", strerror(errno));
		moep_dev_close(dev1);
		moep_dev_close(dev2);
		close(event_all);
		return;
	}

	if (!(radiotap = moep_frame_radiotap(frame)))
		goto out;
	radiotap->hdr.it_version = 0;

	if (!(hdr = moep_frame_moep80211_hdr(frame)))
		goto out;
	hdr->frame_control = htole16(IEEE80211_FTYPE_DATA | IEEE80211_STYPE_DATA);
	memcpy(hdr->ra, "\xde\xad\xbe\xef\x02\x01", sizeof(hdr->ra));
	memcpy(hdr->ta, "\xde\xad\xbe\xef\x02\x02", sizeof(hdr->ta));

	if (!moep_frame_set_moep_hdr_ext(frame, &data_rate_hdr_pctrl.hdr))
		goto out;

	if (!moep_frame_set_payload(frame, data_rate_work_payload,
				    sizeof(data_rate_work_payload)))
		goto out;

	for (k = 0; k < 10; k++) {
		for (i = 0; i < ARRAY_SIZE(legacy_rates); i++) {
			radiotap->rate = legacy_rates[i];
			radiotap->hdr.it_present = BIT(IEEE80211_RADIOTAP_RATE);
			if (moep_dev_tx(dev1, frame)) {
				printf("transmitting packet with data rate %f "
				       "MBit/s failed: %s\n",
				       (float)legacy_rates[i] / 2,
				       strerror(errno));
			};
		}
		for (i = 0; i < 32; i++) {
			radiotap->mcs.mcs = i;
			radiotap->mcs.flags = IEEE80211_RADIOTAP_MCS_BW_20
					   | IEEE80211_RADIOTAP_MCS_SGI;
			radiotap->mcs.known = IEEE80211_RADIOTAP_MCS_HAVE_MCS
					   | IEEE80211_RADIOTAP_MCS_HAVE_BW
					   | IEEE80211_RADIOTAP_MCS_HAVE_GI;
			radiotap->hdr.it_present = BIT(IEEE80211_RADIOTAP_MCS);
			if (moep_dev_tx(dev1, frame)) {
				printf("transmitting packet with MCS index %d "
				       "failed: %s\n", i, strerror(errno));
			}
		}
	}

	timeout.tv_sec = 1;
	timeout.tv_nsec = 0;

	legacy_rate_worked = 0;
	mcs_rate_worked = 0;

	moep_select(0, NULL, NULL, NULL, &timeout, NULL);

	for (i = 0; i < ARRAY_SIZE(legacy_rates); i++) {
		if (legacy_rate_worked & BIT(i)) {
			printf("+");
		} else {
			printf("-");
		}
	}
	printf("|");
	for (i = 0; i < 32; i++) {
		if (mcs_rate_worked & BIT(i)) {
			printf("+");
		} else {
			printf("-");
		}
	}
	printf("\n");

out:
	moep_frame_destroy(frame);
	moep_dev_close(dev1);
	moep_dev_close(dev2);
	close(event_all);
}

int main(int argc, char **argv)
{
	struct arguments args;
	u64 chan1, chan2;
	int i;

	memset(&args, 0, sizeof(args));
	argp_parse(&argp, argc, argv, 0, 0, &args);

	if (dev_avail(args.if1))
		return -1;
	if (dev_avail(args.if2))
		return -1;

	printf("Computing maximum MTU...\n");
	max_mtu(args.if1);
	max_mtu(args.if2);

	printf("Testing available channels...\n");
	chan1 = avail_channels(args.if1);
	chan2 = avail_channels(args.if2);

	printf("Testing working rates...\n");
	for (i = 0; i < ARRAY_SIZE(chan_freq); i++) {
		if (chan1 & chan2 & BIT(i))
			test_data_rates_work(args.if1, args.if2, chan_freq[i]);
	}
	for (i = 0; i < ARRAY_SIZE(chan_freq); i++) {
		if (chan1 & chan2 & BIT(i))
			test_data_rates_work(args.if2, args.if1, chan_freq[i]);
	}

	return 0;
}
